samtranslator/model/sam_resources.py

```diff
@@ -2407,3 +2418,148 @@ def _set_resolver_code_settings(resolver_code_settings: aws_serverless_graphqlap
+            func_config.DataSourceName = self._parse_datasource_name(relative_id, function, datasources)
+            func_config.MaxBatchSize = cast(PassThrough, function.MaxBatchSize)
+            func_config.Description = cast(PassThrough, function.Description)
+
+            if function.Sync:
+                func_config.SyncConfig = self._parse_function_sync(function.Sync)
+
+            if function.Runtime:
+                func_config.Runtime = {
+                    "Name": cast(PassThrough, function.Runtime.Name),
+                    "RuntimeVersion": cast(PassThrough, function.Runtime.Version),
+                }
+
+            resources.append(func_config)
+
+        return resources
+
+    def _check_and_construct_none_datasource(
+        self, functions: Dict[str, aws_serverless_graphqlapi.Function], api_id: Intrinsicable[str]
+    ) -> Optional[DataSource]:
+        """
+        Checks if a DataSource with type "NONE" needs to be created.
+
+        Within a Serverless::GraphQLApi resource, we only generate one "NoneDataSource"
+        that functions can use. This function checks to see if the customer is using a
+        NoneDataSource and constructs it and returns. If there is no functions that use
+        a NoneDataSource, then we return None.
+        """
+        for function in list(functions.values()):
+            if function.Name == "NONE":
+                logical_id = f"{self.logical_id}NoneDataSource"
+
+                none_datasource = DataSource(
+                    logical_id=logical_id, depends_on=self.depends_on, attributes=self.resource_attributes
+                )
+                none_datasource.ApiId = api_id
+                none_datasource.Name = logical_id
+                none_datasource.Type = "NONE"
+
+                return none_datasource
+
+        return None
+
+    def _parse_datasource_name(
+        self,
+        relative_id: str,
+        function: aws_serverless_graphqlapi.Function,
+        datasources: Optional[Dict[str, aws_serverless_graphqlapi.DynamoDBDataSource]],
+    ) -> str:
+        """
+        Parse DataSource name from a Serverless::GraphQLApi function.
+
+        There are 3 different cases for the DataSource name:
+
+        1. Customer defines the "DataSourceName" property, and we can simply return this value.
+
+        2. Customer defines "DataSource" propery as "NONE", so we return the name of the NoneDataSource
+           for this Serverless::GraphQLApi resource.
+
+        3. Customer defines "DataSource" property with a logical id of a datasource defined in
+           Serverless::GraphQLApi. We can then search if this DataSource exists, and return the name.
+           If it does not exist, throw an InvalidResourceException.
+        """
+        if function.DataSourceName:
+            return function.DataSourceName
+
+        if function.DataSource == "NONE":
+            return f"{self.logical_id}NoneDataSource"
```

Comment:
I'd put this as a instance constant or such; I had to read twice as I was confused why is a logical ID returned from a function that mentions the name. And then I noticed this is reused in `_check_and_construct_none_datasource`.